Refactor AimAnimation traits

reaperrr opened the pull request:
Splits `WithAimAnimation` from `WithAttackAnimation` and refactors how/when the aim sequence + prefix are updated.

Edit: Now also removes `ReloadPrefix` from the AimAnimation traits, added a `GrantConditionOnArmamentReloading` trait instead. Switching out the entire sprite body/turret is more reliable than doing these prefix gymnastics.
Supersedes #15078.

Prerequisite for my upcoming animation priority system that aims to fix/avoid all similar and future animation conflicts.

reaperrr committed:
Add INotifyAiming interface

And trigger notifications from Attack* traits.

reaperrr committed:
Add ReloadingCondition to Armament

Allows to grant condition while reloading.

reaperrr commented:
Fixed.

reaperrr committed:
Use Armament.ReloadingCondition on RA V2

The new property makes the AmmoPool work-around obsolete.

reaperrr commented:
Updated: Now also removes `ReloadPrefix` from the `*AimAnimation` traits, added a `GrantConditionOnArmamentReloading` trait instead. Switching out the entire sprite body/turret is more reliable than doing these prefix gymnastics.

GraionDilach commented on OpenRA.Mods.Common/Traits/Conditions/GrantConditionOnArmamentReloading.cs:
```diff
@@ -0,0 +1,83 @@
+			: base(info)
+		{
+			armament = self.TraitsImplementing<Armament>()
+				.Single(a => a.Info.Name == Info.Armament);
```
Do we have an official policy on this one? How should eliteweapons be treated in this case? How are they treated in TS?

reaperrr replied:
> How should eliteweapons be treated in this case?

Elite weapons should have a different name than their non-elite counterpart, then you can use a 2nd GCOAR for the elite weapon.

@pchote said months ago that we should ultimately aim for one attack trait per armament, so each armament having a distinct name will become mandatory anyway, eventually.
I'm trying to write anything somewhat related with that in mind already, to keep the number of places that need to be adapted later as low as possible.

> How are they treated in TS?

Probably still using "primary" for both, but iirc there aren't any actors in TS that have both an elite weapon *and* one of the affected traits, so at this point it shouldn't matter.

pchote replied:
> @pchote said months ago that we should ultimately aim for one attack trait per armament, so each armament having a distinct name will become mandatory anyway, eventually.

Hmm, I don't remember this.

reaperrr replied:
Ok... anyway, I think the approach of this PR is much more robust and less issue-prone than attempting to control every Armament via a single condition trait would be.

However, it just occured to me that maybe making the Armament itself grant a `ReloadingCondition` would be even better. What do you think?

abcdefg30 commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -51,36 +45,32 @@ public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
 		{
-			if (!IsTraitDisabled && !wsb.IsTraitDisabled && !string.IsNullOrEmpty(Info.AttackSequence))
+			if (!IsTraitDisabled && !wsb.IsTraitDisabled && !string.IsNullOrEmpty(Info.Sequence))
 			{
```
Braces around a single-line statement.

abcdefg30 commented on OpenRA.Mods.Common/UpdateRules/Rules/RemoveReloadPrefix.cs:
```diff
@@ -0,0 +1,73 @@
+		{
+			get
+			{
+				return "ReloadPrefix has been removed from WithAimAnimation and WithTurretAimAnimation.\n"
```
Am I mixing something up here or did you just add `WithAimAnimation` in this PR together with an update rule refactoring the new trait?

reaperrr replied:
Hm yeah, now that you mention it... guess I should put all the WithAimAnimation stuff in one update rule and let this here only update WithTurretAimAnimation.

pchote replied:
The simpler the better.  What is the strategy forward here?

pchote commented:
This is a dependency for #15008 and ties in with the various update rule messyness, so adding to the milestone.

reaperrr committed:
Refactor WithTurretAimAnimation using INotifyAiming

Triggering start and end of the aiming animation via interface, as well as removing ReloadPrefix (in favor of switching sprite bodies via condition when reloading), allows us to drop updating via ITick, which in turn will make it much easier to ultimately make this trait compatible with other animation traits, once the priority system lands.

reaperrr committed:
Split and refactor aim animation from WithAttackAnimation

Splitting it from the attack animation, triggering start and end of the aiming animation via interface, as well as removing ReloadPrefix (in favor of switching sprite bodies via condition when reloading) allows us to drop updating via ITick, which in turn will make it much easier to ultimately make this trait compatible with other animation traits, once the priority system lands.

reaperrr committed:
Update rule for AimAnimation splits and ReloadPrefix removal

reaperrr commented:
Updated:
- merged #15139 back into this
- merged all update rules into one
- reorganised and restructured commits to make the whole thing easier to review (and update, if necessary)

The only point I haven't adressed yet, since I wanted to get some feedback first:
> However, it just occured to me that maybe making the Armament itself grant a `ReloadingCondition` (instead of adding `GrantConditionOnArmamentReloading`) would be even better. What do you think?

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -61,6 +61,10 @@ public class ArmamentInfo : PausableConditionalTraitInfo, Requires<AttackBaseInf
 
+		[GrantedConditionReference]
+		[Desc("Condition to grant while reloading.")]
+		public readonly string OnReloadingCondition = null;
```
IMO this should be `ReloadingCondition` for consistency.  We don't do e.g. `OnAirborneCondition` or `OnCruisingCondition` in `Aircraft`.

reaperrr replied:
I was a bit unsure about this, because - purely from a wording perspective - some modders might misinterpret ReloadingCondition as "this condition is needed *for* reloading".
On the other hand, that's what the Desc is for, and modders should know from other traits that the property name would include "Requires" if that were its purpose.

pchote reviewed (commented):
The code changes look good to me, but I haven't tested this yet.  In the meantime, I do have one minor nit/opinion for discussion:

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAimAnimation.cs:
```diff
@@ -0,0 +1,75 @@
+
+		public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
+		{
+			var matches = ai.TraitInfos<WithSpriteBodyInfo>().Count(w => w.Name == Body);
```
Style nit: `.Count` requires walking the entire enumeration, which is wasteful in cases where we don't need to know what the actual count is.  When we want to know if there is exactly one it is conceptually better to use `SingleOrDefault` and then check for null.  In practice this will usually walk the entire enumeration too, but it can return early as soon as it finds a second element that matches the condition.

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAimAnimation.cs:
```diff
@@ -0,0 +1,75 @@
+
+		void INotifyAiming.StartedAiming(Actor self, AttackBase ab)
+		{
+			wsb.DefaultAnimation.ReplaceAnim(Sequence);
```
IMO the code would be clearer if we could do:

```csharp
protected void UpdateSequence()
{
    var seq = !IsTraitDisabled && attack.IsAiming ? Info.Sequence : wsb.Info.Sequence;
    wsb.DefaultAnimation.ReplaceAnim(seq);
}

void INotifyAiming.StartedAiming(Actor self, AttackBase ab) { UpdateSequence(); }
void INotifyAiming.StoppedAiming(Actor self, AttackBase ab) { UpdateSequence(); }
protected override void TraitEnabled(Actor self) { UpdateSequence(); }
protected override void TraitDisabled(Actor self) { UpdateSequence(); }
```

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -51,36 +45,30 @@ public override void RulesetLoaded(Ruleset rules, ActorInfo ai)
 
 		void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barrel)
 		{
+			if (a != armament)
```
This can be merged into the check below: `if (a == armament && Info.DelayRelativeTo == AttackDelayType.Attack)`

pchote commented on OpenRA.Mods.Common/Traits/Render/WithAttackAnimation.cs:
```diff
@@ -92,6 +80,9 @@ void INotifyAttack.Attacking(Actor self, Target target, Armament a, Barrel barre
 
 		void INotifyAttack.PreparingAttack(Actor self, Target target, Armament a, Barrel barrel)
 		{
+			if (a != armament)
```
Likewise here

pchote commented on OpenRA.Mods.Common/Traits/Render/WithTurretAimAnimation.cs:
```diff
@@ -23,57 +23,45 @@ public class WithTurretAimAnimationInfo : ConditionalTraitInfo, Requires<WithSpr
 		readonly WithSpriteTurret wst;
-		string sequence;
+
+		protected string Sequence { get { return !IsTraitDisabled && attack.IsAiming ? Info.Sequence : wst.Info.Sequence; } }
```
Can do the same simplification here as in `WithAimAnimation`.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/SplitAimAnimations.cs:
```diff
@@ -0,0 +1,185 @@
+		public override IEnumerable<string> AfterUpdate(ModData modData)
+		{
+			var message1 = "AimSequences have been split to With*AimAnimation.\n"
+				+ "The following locations have been updated and might need manual adjustments:\n"
```
Can you please change "locations" to "actors" here?

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/SplitAimAnimations.cs:
```diff
@@ -0,0 +1,185 @@
+
+			var message2 = "ReloadPrefixes have been removed.\n"
+				+ "Instead, grant a condition on reloading via Armament.ReloadingCondition to enable\n"
+				+ "an alternate sprite body (with reloading sequences) at the following locations:\n"
```
likewise here "on the following actors"

pchote commented on OpenRA.Mods.Common/Traits/Armament.cs:
```diff
@@ -169,8 +177,24 @@ protected override void Created(Actor self)
 
+		void UpdateCondition(Actor self)
+		{
+			if (conditionManager == null)
```
`if (string.IsNullOrEmpty(Info.ReloadingCondition) || conditionManager == null)`!!!

Right now we crash when anything except the V2 fires its weapon.

pchote reviewed (dismissed):
LGTM :+1:.  Just a few minor nits and one major issue that has an easy fix.

pchote commented:
The `ReloadingCondition` and `INotifyAiming` are going to be really useful going forwards.  If #15008 goes ahead with a `GrantConditionWhileAiming` trait then we may be able to reproduce the original C&C rocket launcher behaviour where it stops moving and unlocks its turret while aiming.

reaperrr commented:
Updated.

pchote commented on OpenRA.Mods.Common/UpdateRules/Rules/SplitAimAnimations.cs:
```diff
@@ -0,0 +1,185 @@
+
+			var message2 = "ReloadPrefixes have been removed.\n"
+				+ "Instead, grant a condition on reloading via Armament.ReloadingCondition to enable\n"
+				+ "an alternate sprite body (with reloading sequences) at the following actors:\n"
```
at &rarr; on.

reaperrr commented:
Fixed.

pchote commented:
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/8e9ca46e1ac0f41d8986599ab4878ef5eca59eb5

pchote commented:
#15201 reveals that this PR didn't rename `AttackSequence` &rarr; `Sequence` in TS.  Did you forget to update/stage these changes, or was this a failing in the update rule?

reaperrr commented:
I just forgot to update TS.
